src/sentry/replays/usecases/delete.py

```diff
@@ -0,0 +1,155 @@
+    publisher = initialize_replays_publisher(is_async=True)
+    for replay_id in replay_ids:
+        publisher.publish("ingest-replay-events", archive_event(project_id, replay_id))
+    publisher.flush()
+
+
+def delete_replay_recordings(project_id: int, row: MatchedRow) -> None:
+    with cf.ThreadPoolExecutor(max_workers=100) as pool:
+        pool.map(_delete_if_exists, _make_recording_filenames(project_id, row))
+
+
+def _delete_if_exists(filename: str) -> None:
+    """Delete the blob if it exists or silence the 404."""
+    try:
+        storage_kv.delete(filename)
+    except NotFound:
+        pass
+
+
+def _make_recording_filenames(project_id: int, row: MatchedRow) -> list[str]:
+    # We assume every segment between 0 and the max_segment_id exists. Its a waste of time to
+    # delete a non-existent segment but its not so significant that we'd want to query ClickHouse
+    # to verify it exists.
+    replay_id = row["replay_id"]
+    retention_days = row["retention_days"]
+    platform = row["platform"]
+
+    filenames = []
+    for segment_id in range(row["max_segment_id"] + 1):
+        segment = RecordingSegmentStorageMeta(project_id, replay_id, segment_id, retention_days)
+        filenames.append(make_recording_filename(segment))
+        if platform != "javascript":
+            filenames.append(make_video_filename(segment))
+
+    return filenames
+
+
+class MatchedRow(TypedDict):
+    retention_days: int
+    replay_id: str
+    max_segment_id: int
+    platform: str
+
+
+class MatchedRows(TypedDict):
+    rows: list[MatchedRow]
+    has_more: bool
+
+
+def fetch_rows_matching_pattern(
+    project_id: int,
+    start: datetime,
+    end: datetime,
+    query: str,
+    environment: list[str],
+    limit: int,
+    offset: int,
+) -> MatchedRows:
+    search_filters = parse_search_query(query, config=replay_url_parser_config)
+    where = handle_search_filters(scalar_search_config, search_filters)
+
+    if environment:
+        where.append(Condition(Column("environment"), Op.IN, environment))
+
+    query = Query(
+        match=Entity("replays"),
+        select=[
+            Function("any", parameters=[Column("retention_days")], alias="retention_days"),
+            Column("replay_id"),
+            Function("max", parameters=[Column("segment_id")], alias="max_segment_id"),
+            Function("any", parameters=[Column("platform")], alias="platform"),
+        ],
+        where=[
+            Condition(Column("project_id"), Op.EQ, project_id),
+            Condition(Column("timestamp"), Op.LT, end),
+            Condition(Column("timestamp"), Op.GTE, start),
+            *where,
+        ],
+        groupby=[Column("replay_id")],
+        orderby=[OrderBy(Function("min", parameters=[Column("timestamp")]), Direction.ASC)],
+        granularity=Granularity(3600),
+        limit=Limit(limit),
+        offset=Offset(offset),
+    )
+
+    response = execute_query(
```

Comment:
if you end up needing more robust retrying, you can check out `_retry_operation` https://github.com/getsentry/sentry/blob/31b8a0c441d69fd8a9dd57c13b2e7f63a2a082df/src/sentry/tasks/embeddings_grouping/utils.py#L349, as we had many issues with snuba rate limits in seer backfills. you may not have issues here as it's not as parallelized

Reply:
Good idea.  That reminds me we have similar code here which may be compatible for both our use cases: https://github.com/getsentry/sentry/blob/fa61b051268501c828fd1a8aacdc117260e700c7/src/sentry/filestore/gcs.py#L411-L414